### ISSUE TITLE
<string>: basic_string's operator+ does strlen on the input one too many times

### ISSUE THREAD

**Issue by BillyONeal** [microsoft/STL#456]
While reviewing #419 I noticed something strange about how we handle `operator+`. Consider the following:

https://github.com/microsoft/STL/blob/c5cde6ecbaa661fd23ad085e9e716e843059b104/stl/inc/xstring#L4279-L4289

Here, we do the obvious `strlen` (`_Traits::length`) on line 4285 to figure out the length of the `const _Elem*` parameter. However, then we call `operator+=(const _Elem*)`, which is going to do *another* `_Traits::length` on that input.

This should be reworked to call `_Traits::length` only once, and ideally call an internal-only helper version of append that avoids creating extra exception handling states (since we know the appends here cannot reallocate, but the compiler does not know that).

**Comment by BillyONeal** [comment]
(I'm thinking something like this pseudocode but I've not spent enough time on it to prove it correct:

```
basic_string(_Concat_strings_tag_t, const _Alloc &_Al, const _Elem *const _Left_ptr, const size_type _Left_size,
             const _Elem *const _Right_ptr, const size_type _Right_size)
    : _Mypair(_One_then_variadic_args_t(), _Alty_traits::select_on_container_copy_construction(_Al)) {
    _STL_INTERNAL_CHECK(_Elem() == _Right_ptr[_Right_size]);
    auto &&_Alproxy = _GET_PROXY_ALLOCATOR(_Alty, _Getal());
    _Container_proxy_ptr<_Alty> _Proxy(_Alproxy, _Mypair._Myval2);
    _Tidy_init();
    _Reallocate_for(
        _Left_size + _Right_size,
        [](_Elem *const _New_ptr, const _Elem *const _Left_ptr, const size_type _Left_size,
           const _Elem *const _Right_ptr, const size_type _Right_size) noexcept {
            _Traits::copy(_New_ptr, _Left_ptr, _Left_size);
            _Traits::copy(_New_ptr + _Left_size, _Right_ptr, _Right_size + 1);
        },
        _Left_ptr, _Left_size, _Right_ptr, _Right_size);
    _Proxy._Release();
}

// later

template <class _Elem, class _Traits, class _Alloc>
_NODISCARD basic_string<_Elem, _Traits, _Alloc> operator+(
    const basic_string<_Elem, _Traits, _Alloc>& _Left, const basic_string<_Elem, _Traits, _Alloc>& _Right) {
    return basic_string<_Elem, _Traits, _Alloc>(_Concat_strings_tag, _Left._Getal(),
        _Left.data(), _Left.size(), _Right.data(), _Right.size());
}

template <class _Elem, class _Traits, class _Alloc>
_NODISCARD basic_string<_Elem, _Traits, _Alloc> operator+(
    const _Elem* const _Left, const basic_string<_Elem, _Traits, _Alloc>& _Right) {
    return basic_string<_Elem, _Traits, _Alloc>(_Concat_strings_tag, _Left._Getal(),
        _Left, _Convert_size<_Size_type>(_Traits::length(_Left)), _Right.data(), _Right.size());
}

template <class _Elem, class _Traits, class _Alloc>
_NODISCARD basic_string<_Elem, _Traits, _Alloc> operator+(
    const basic_string<_Elem, _Traits, _Alloc>& _Left, const _Elem* const _Right) {
    return basic_string<_Elem, _Traits, _Alloc>(_Concat_strings_tag, _Left._Getal(),
        _Left.data(), _Left.size(). _Right, _Convert_size<_Size_type>(_Traits::length_Right)));
}
```

**Comment by BillyONeal** [comment]
Also probably best to avoid touching this until #419 is merged or disposed of (or fix #53 in the same PR) to avoid so many merge conflicts.
